playgame.py

```diff
@@ -5,55 +5,111 @@
+import os
 from tools.hanabi_moves import \
     HanabiDiscardAction, \
     HanabiPlayAction, \
     HanabiColorDiscloseAction, \
     HanabiRankDiscloseAction
 
-logger = logging.getLogger(__name__)
-
 def main(argv):
     args = parse_args()
-    prep_logger(args.log_dir, args.verbose, args.log_stderr)
+    logger = prep_logger(args.log_dir, args.verbose, args.log_stderr, len(args.players))
+    if args.is_tournament:
+        run_tournament(args, logger)
+    else:
+        run_one_game(args, logger)
+
+def generate_pairings(players):
+    res = []
+    for i in range(len(players)):
+        for j in range(i + 1, len(players)):
+            res.append([players[i], players[j]])
+    return res
+
+def run_tournament(args, logger):
+    tournament_scores = dict.fromkeys(args.players, 0)
+    pairings = generate_pairings(args.players)
+    for player1, player2 in pairings:
+        try:
+            game = HanabiGame([player1, player2], args.seed, args.variant)
+            game.play_game(args, logger)
+            score = game.table.score()
+        except:
+            #if a player has messed up, penalize both
+            #could diqualify the failing bot if we can determine who failed
+            logger.warning("{player1} or {player2} failed to complete a game".format(
+                player1 = player1,
+                player2 = player2
+                ))
+            score = -25
+        finally:
+            for handler in logger.handlers:
+                if handler.__class__ is RotatingFileHandler:
+                    handler.doRollover()
+            tournament_scores[player1] += score
+            tournament_scores[player2] += score
+    tournament_scores = {k: v / len(pairings) for k, v in tournament_scores.iteritems()}
+    logger.info("Scores: {scores}".format(scores = tournament_scores))
+    winning_score = max(tournament_scores.itervalues())
+    winners = [key for key, value in tournament_scores.items() if value == winning_score]
+    if len(winners) == 1:
+        logger.info('Winner is {winner}'.format(winner = winners[0]))
+    else:
+        logger.info('Winners are {winners}'.format(winners = winners))
+
+def run_one_game(args, logger):
     game = HanabiGame(args.players, args.seed, args.variant)
-    game.play_game(args)
+    game.play_game(args, logger)
 
-def prep_logger(log_dir, verbose, log_stderr):
+def prep_logger(log_dir, verbose, log_stderr, count):
+    logging.basicConfig()
+    logger = logging.getLogger(__name__)
+    formatter = logging.Formatter('[%(asctime)s] %(levelname)8s --- %(message)s ' +
+                                  '(%(filename)s:%(lineno)s)',datefmt='%Y-%m-%d %H:%M:%S')
     if verbose:
+        logger.setLevel(logging.DEBUG)
+    else:
         logger.setLevel(logging.INFO)
-    logger.addHandler(logging.StreamHandler())
     if log_dir:
-        fh = logging.FileHandler(log_dir)
+        fh = RotatingFileHandler('results/{dir}'.format(dir = log_dir), mode = 'w', backupCount = count)
```

Comment:
Will this automatically create the results directory if one does not exist?

Reply:
just tested, it fails to run.
time to import os again!

Here's an interesting question, should we implicitly save to results, or still require the script to call it out? option 1 makes sure the directories are clean. Option 2 avoids confusing unintended behavior

Reply:
Opened #26